# Dev

+ script to read bams, output normalized counts table

#Bennett

+ high-vquest of joined reads

#Deterding

+ inquire about email

#Hits-Clip

+ peaks through hits-clip coming up as 60 base pairs

## tracks for PK61, PK62, PK63

+ still need to run PK61-63
+ heatmap of miRNA, something that you can zoom into
+ matrix2png

#Poly(A)

+ normal v tumor is not being observed as differentially significant after dexseq
+ unadjusted pvalues; cluster those genes through kegg
+ for peter's, likely will want to use pvalue rather and padjust

+ intronic peaks; start and stop of gene model as separate run; let austin know its location
+ don't necessarily test with these sites, but may want to add yet another track with all sites + intronic peaks

+ map to mouse, then to human for peter's previous samples (samples with an 'a')
    + map some human to mouse and see how many map (percentage wise)
    + possible to mask human genome from mouse genome based on a certain length

    + testing mouse -- for each polya grab 25, 50, 75, 100 from the site
        + align to mouse and see what maps for each length
        + are there gains in using longer length reads

#Spillman; Human; ChIP

+ Minus
    + E2Input_minus
    + E21PR1_minus
    + E21PR2_minus
    + E21PR3_minus
+ Plus
    + E2input_plus
    + E21PR1_plus
    + E21PR2_plus
    + E21PR3_plus

#Thaikoottathil

+ run analysis through Homer; annotate

#Walter

+ levin classification; score
+ given list with direction, compute a score for each patient along with their group (TB, noinf, etc.)
+ calculate threshold for plot

+ adding total intensity at up-regulated transcripts and subtracting total intensity at all down-regulated transcripts
