<html>
<body>

<div>
<h1>Description</h1>
<p>Dawn Duval; canine exome sequencing project</p>
</div>
<div>
	<h1>
		Display Conventions
	</h1>
	<p>
		Each track is colored according to the following legend
	</p>
	<table>
		<thead>
			<tr>
				<th>
					Mark
				</th>
				<th>
					Color
				</th>
			</tr>
		</thead>
		<tbody>
			<tr>
				<td>
					1025
				</td>
				<td style="background-color: rgb(35, 139, 69);">
					&nbsp;
				</td>
			</tr>
			<tr>
				<td>
					113
				</td>
				<td style="background-color: rgb(136, 65, 157);">
					&nbsp;
				</td>
			</tr>
			<tr>
				<td>
					166408
				</td>
				<td style="background-color: rgb(43, 140, 190);">
					&nbsp;
				</td>
			</tr>
			<tr>
				<td>
					172160
				</td>
				<td style="background-color: rgb(215, 48, 31);">
					&nbsp;
				</td>
			</tr>
			<tr>
				<td>
					22
				</td>
				<td style="background-color: rgb(5, 11, 176);">
					&nbsp;
				</td>
			</tr>
			<tr>
				<td>
					353GAAACC
				</td>
				<td style="background-color: rgb(2, 129, 138);">
					&nbsp;
				</td>
			</tr>
			<tr>
				<td>
					353GGCTAC
				</td>
				<td style="background-color: rgb(206, 18, 86);">
					&nbsp;
				</td>
			</tr>
			<tr>
				<td>
					36
				</td>
				<td style="background-color: rgb(174,1,126);">
					&nbsp;
				</td>
			</tr>
			<tr>
				<td>
					400
				</td>
				<td style="background-color: rgb(35,132,67);">
					&nbsp;
				</td>
			</tr>
			<tr>
				<td>
					500
				</td>
				<td style="background-color: rgb(34,94,168);">
					&nbsp;
				</td>
			</tr>
			<tr>
				<td>
					522
				</td>
				<td style="background-color: rgb(204,76,2);">
					&nbsp;
				</td>
			</tr>
			<tr>
				<td>
					730
				</td>
				<td style="background-color: rgb(227,26,28);">
					&nbsp;
				</td>
			</tr>
			<tr>
				<td>
					735
				</td>
				<td style="background-color: rgb(102,194,164);">
					&nbsp;
				</td>
			</tr>
			<tr>
				<td>
					868
				</td>
				<td style="background-color: rgb(140,150,198);">
					&nbsp;
				</td>
			</tr>
			<tr>
				<td>
					Bililey
				</td>
				<td style="background-color: rgb(123,204,196);">
					&nbsp;
				</td>
			</tr>
			<tr>
				<td>
					KTCC
				</td>
				<td style="background-color: rgb(252,141,89);">
					&nbsp;
				</td>
			</tr>
		</tbody>
	</table>
    <h1>Mapping Results</h1>
    <table border="1" class="dataframe">
      <thead>
        <tr style="text-align: right;">
          <th></th>
          <th>Paired Reads</th>
          <th>Pairs Aligned</th>
          <th>Read Sequences</th>
          <th>Aligned</th>
          <th>Unique Alignment</th>
          <th>Gapped Alignment</th>
          <th>Quality Filter</th>
          <th>Homopolymer Filter</th>
          <th>Perc Aligned</th>
          <th>Perc Uniquely Aligned</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th>1025</th>
          <td> 34617873</td>
          <td> 33505312</td>
          <td> 69235746</td>
          <td> 68970324</td>
          <td> 64642802</td>
          <td> 3965372</td>
          <td> 130032</td>
          <td> 4606</td>
          <td> 99.61664</td>
          <td> 93.36622</td>
        </tr>
        <tr>
          <th>113</th>
          <td> 28706841</td>
          <td> 27160629</td>
          <td> 57413682</td>
          <td> 56998776</td>
          <td> 52633839</td>
          <td> 5750490</td>
          <td> 147024</td>
          <td> 5374</td>
          <td> 99.27734</td>
          <td> 91.67473</td>
        </tr>
        <tr>
          <th>166408</th>
          <td> 49149698</td>
          <td> 46977972</td>
          <td> 98299396</td>
          <td> 97865897</td>
          <td> 91362279</td>
          <td> 7400935</td>
          <td> 206105</td>
          <td> 7974</td>
          <td>   99.559</td>
          <td> 92.94287</td>
        </tr>
        <tr>
          <th>172160</th>
          <td> 46962026</td>
          <td> 46367317</td>
          <td> 93924052</td>
          <td> 93559913</td>
          <td> 89106862</td>
          <td> 2754920</td>
          <td> 151813</td>
          <td> 4513</td>
          <td>  99.6123</td>
          <td> 94.87119</td>
        </tr>
        <tr>
          <th>22</th>
          <td> 15734738</td>
          <td> 15235468</td>
          <td> 31469476</td>
          <td> 31317385</td>
          <td> 29075470</td>
          <td> 1638850</td>
          <td>  64084</td>
          <td> 2156</td>
          <td>  99.5167</td>
          <td> 92.39261</td>
        </tr>
        <tr>
          <th>353GAAACC</th>
          <td> 46001933</td>
          <td> 45490636</td>
          <td> 92003866</td>
          <td> 91698422</td>
          <td> 87426112</td>
          <td> 2484641</td>
          <td> 132515</td>
          <td> 4349</td>
          <td> 99.66801</td>
          <td> 95.02439</td>
        </tr>
        <tr>
          <th>353GGCTAC</th>
          <td> 17964266</td>
          <td> 16945021</td>
          <td> 35928532</td>
          <td> 35691511</td>
          <td> 33037742</td>
          <td> 3666892</td>
          <td>  99448</td>
          <td> 3697</td>
          <td>  99.3403</td>
          <td> 91.95405</td>
        </tr>
        <tr>
          <th>36</th>
          <td> 26045804</td>
          <td> 25200655</td>
          <td> 52091608</td>
          <td> 51896818</td>
          <td> 48563025</td>
          <td> 3183521</td>
          <td>  82552</td>
          <td> 3313</td>
          <td> 99.62606</td>
          <td>  93.2262</td>
        </tr>
        <tr>
          <th>400</th>
          <td> 23190716</td>
          <td> 22388602</td>
          <td> 46381432</td>
          <td> 46134125</td>
          <td> 42995424</td>
          <td> 2483894</td>
          <td> 111516</td>
          <td> 3342</td>
          <td>  99.4668</td>
          <td> 92.69965</td>
        </tr>
        <tr>
          <th>500</th>
          <td> 24869093</td>
          <td> 23585005</td>
          <td> 49738186</td>
          <td> 49326419</td>
          <td> 45814064</td>
          <td> 4267012</td>
          <td> 124617</td>
          <td> 3909</td>
          <td> 99.17213</td>
          <td> 92.11044</td>
        </tr>
        <tr>
          <th>522</th>
          <td> 40335846</td>
          <td> 38310074</td>
          <td> 80671692</td>
          <td> 80231928</td>
          <td> 74412109</td>
          <td> 6229208</td>
          <td> 174839</td>
          <td> 6352</td>
          <td> 99.45487</td>
          <td> 92.24067</td>
        </tr>
        <tr>
          <th>730</th>
          <td> 40691286</td>
          <td> 39703696</td>
          <td> 81382572</td>
          <td> 81102396</td>
          <td> 75974512</td>
          <td> 4232995</td>
          <td> 128788</td>
          <td> 3175</td>
          <td> 99.65573</td>
          <td> 93.35477</td>
        </tr>
        <tr>
          <th>735</th>
          <td> 25915370</td>
          <td> 24194159</td>
          <td> 51830740</td>
          <td> 51530695</td>
          <td> 47724363</td>
          <td> 5116392</td>
          <td> 138156</td>
          <td> 4544</td>
          <td> 99.42111</td>
          <td> 92.07733</td>
        </tr>
        <tr>
          <th>868</th>
          <td> 38665005</td>
          <td> 36714024</td>
          <td> 77330010</td>
          <td> 76944643</td>
          <td> 70605737</td>
          <td> 6016122</td>
          <td> 171556</td>
          <td> 5452</td>
          <td> 99.50166</td>
          <td> 91.30445</td>
        </tr>
        <tr>
          <th>Bililey</th>
          <td> 49555874</td>
          <td> 49033036</td>
          <td> 99111748</td>
          <td> 98745718</td>
          <td> 94022296</td>
          <td> 2418757</td>
          <td> 156158</td>
          <td> 3448</td>
          <td> 99.63069</td>
          <td> 94.86494</td>
        </tr>
        <tr>
          <th>KTCC</th>
          <td> 36839016</td>
          <td> 36547567</td>
          <td> 73678032</td>
          <td> 73344462</td>
          <td> 69518619</td>
          <td> 1483235</td>
          <td>  66971</td>
          <td>  538</td>
          <td> 99.54726</td>
          <td> 94.35461</td>
        </tr>
      </tbody>
    </table>
</div>


<div>
<h1>Methods</h1>

<p>Reads were first concatenated from separate lanes into a single fastq for R1 and R2.</p>

<h2>Mapping</h2>

<pre><code>
    #!/usr/bin/env bash
    #BSUB -J align[1-16]
    #BSUB -e align.%J.%I.err
    #BSUB -o align.%J.%I.out
    #BSUB -q normal
    #BSUB -R "select[mem>16] rusage[mem=16] span[hosts=1]"
    #BSUB -n 8
    #BSUB -P duval

    set -o nounset -o pipefail -o errexit -x
    source $HOME/projects/duval_exome/bin/config.sh

    sample=${SAMPLES[$(($LSB_JOBINDEX - 1))]}
    r1=$DATA/${sample}_R1.fastq.gz
    r2=$DATA/${sample}_R2.fastq.gz
    results=$RESULTS/$sample
    bam=$results/$sample.bam
    stats=$results/$sample.alignment_stats.txt

    if [[ ! -d $results ]]; then
        mkdir -p $results
    fi

    if [[ ! -f $bam ]]; then
        novoalign -d $NOVOIDX -f $r1 $r2 -o SAM "@RG\tID:$sample\tSM:$sample\tPU:Illumina\tLB:PE" -r None -i 250 100 -c 8 -k \
            2> $stats \
            | samtools view -ShuF4 - \
            | samtools sort -o -m 8G - $sample.temp \
            > $bam
    fi
</code></pre>

<h2>Variant detection</h2>

<pre><code>
    #!/usr/bin/env bash
    #BSUB -J duval_variants[1-16]
    #BSUB -e duval_variants.%J.%I.err
    #BSUB -o duval_variants.%J.%I.out
    #BSUB -q normal
    #BSUB -R "select[mem>24] rusage[mem=24] span[hosts=1]"
    #BSUB -n 1
    #BSUB -P duval_exome

    set -o nounset -o pipefail -o errexit -x
    source $HOME/projects/duval_exome/bin/config.sh

    sample=${SAMPLES[$(($LSB_JOBINDEX - 1))]}

    java='java -Xmx24g -jar'
    bam=$RESULTS/$sample/$sample.bam
    rgbam=${bam/.bam/.rg.bam}
    nodups=${bam/.bam/.nodups.bam}
    reordered=${bam/.bam/.reordered.bam}
    duplicatemetrics=${bam/.bam/.dup_metrics.txt}
    targetintervals=${bam/.bam/.intervals}
    realigned=${bam/.bam/.realign.bam}
    vcf=${bam/.bam/.vcf}

    if [[ -f $vcf ]]; then
        echo "processing complete for $sample"
        exit 0
    fi

    # picard
    READ_GROUP_OPTS="INPUT=$bam OUTPUT=$rgbam SORT_ORDER=coordinate RGID=$sample RGLB=PE RGPL=illumina RGPU=$sample RGSM=$sample"
    MARK_DUPLICATES_OPTS="ASSUME_SORTED=true INPUT=$rgbam OUTPUT=$nodups METRICS_FILE=$duplicatemetrics CREATE_INDEX=true MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=1000 REMOVE_DUPLICATES=true MAX_RECORDS_IN_RAM=800000"
    REORDER_SAM_OPTS="INPUT=$nodups OUTPUT=$reordered REFERENCE=$REFERENCE CREATE_INDEX=true"

    # gatk
    REALIGNER_TARGET_CREATOR="--analysis_type RealignerTargetCreator --reference_sequence $REFERENCE --input_file $reordered --out $targetintervals"
    INDEL_REALIGNER="--analysis_type IndelRealigner --reference_sequence $REFERENCE --input_file $reordered --targetIntervals $targetintervals --out $realigned"

    if [ -f $bam ] && [ ! -f $rgbam ]; then
        $java $PICARD/AddOrReplaceReadGroups.jar $READ_GROUP_OPTS
    fi
    if [ -f $rgbam ] && [ ! -f $nodups ]; then
        $java $PICARD/MarkDuplicates.jar $MARK_DUPLICATES_OPTS
    fi
    # reorder the bam to match the reference contig order
    if [ -f $nodups ] && [ ! -f $reordered ]; then
        $java $PICARD/ReorderSam.jar $REORDER_SAM_OPTS
    fi
    if [ -f $reordered ] && [ ! -f $targetintervals ]; then
        $java $GATK $REALIGNER_TARGET_CREATOR
    fi
    if [ -f $targetintervals ] && [ ! -f $realigned ]; then
        $java $GATK $INDEL_REALIGNER
    fi
    if [ -f $realigned ] && [ ! -f $vcf ]; then
        freebayes -b $realigned -v $vcf -f $REFERENCE -0
    fi
</code></pre>

<h2>SNP annotation</h2>

<pre><code>
    #!/usr/bin/env bash
    #BSUB -J snpeff[1-16]
    #BSUB -e snpeff.%J.%I.err
    #BSUB -o snpeff.%J.%I.out
    #BSUB -q normal
    #BSUB -R "select[mem>8] rusage[mem=8] span[hosts=1]"
    #BSUB -n 1
    #BSUB -P duval_exome

    set -o nounset -o pipefail -o errexit -x
    source $HOME/projects/duval_exome/bin/config.sh

    sample=${SAMPLES[$(($LSB_JOBINDEX - 1))]}

    vcf=$RESULTS/$sample/$sample.vcf
    snpeff=$RESULTS/$sample/$sample.snpeff.txt.gz

    if [[ ! -f $snpeff ]]; then
        java -jar $SNPEFF eff \
            -chr chr \
            -fi $TARGETS \
            -minC 10 \
            -no-downstream \
            -no-intergenic \
            -no-intron \
            -no-upstream \
            -noStats \
            -v \
            -c $SNPEFFCONFIG \
            -o txt \
            canFam3 $vcf \
            | gzip -c > $snpeff
    fi
</code></pre>

<h2>Post-processing</h2>

<pre><code>
    #!/usr/bin/env bash
    #BSUB -J postprocessing
    #BSUB -e postprocessing.%J.err
    #BSUB -o postprocessing.%J.out
    #BSUB -q normal
    #BSUB -R "select[mem>4] rusage[mem=4] span[hosts=1]"
    #BSUB -n 1
    #BSUB -P duval_exome

    <<DOC
    overlap the cancer samples and filter down to most putative mutations

    Tumor   Normal
    500     172160
    353     353M2
    400     166408

    Sample 353 is tagged GGCTAC
    Sample 353M2 is tagged GAAACC
    DOC

    set -o nounset -o pipefail -o errexit -x
    source $HOME/projects/duval_exome/bin/config.sh

    intersect_results=$RESULTS/intersect_results
    if [[ ! -d $intersect_results ]]; then
        mkdir -p $intersect_results
    fi

    normal_samples=(172160 353GAAACC 166408)
    # the first three are paired with normals
    tumor_samples=(500 353GGCTAC 400 1025 113 22 36 522 730 735 868 Bililey KTCC)

    # run normal/cancer pairs with filtering
    for (( i = 0; i < ${#normal_samples[@]}; i++ )); do
        normal=${normal_samples[$i]}
        tumor=${tumor_samples[$i]}
        vcf=$intersect_results/filtering_with_pair_only/$tumor.vcf
        snpeff=$intersect_results/filtering_with_pair_only/$tumor.snpeff.txt.gz
        annotated=$intersect_results/filtering_with_pair_only/$tumor.annotated_snps.txt.gz
        if [[ ! -f $vcf ]]; then
            bedtools intersect -v -a $RESULTS/$tumor/$tumor.freebayes.vcf -b $RESULTS/$normal/$normal.freebayes.vcf > $vcf
        fi
        if [[ ! -f $snpeff ]]; then
            java -jar $SNPEFF eff -chr chr -noStats -v -c $SNPEFFCONFIG \
                -fi $TARGETS -o txt -minQ 10 -minC 10 -no-downstream -no-intergenic \
                -no-intron -no-upstream canFam3 $vcf | gzip -c > $snpeff
        fi
        if [[ ! -f $annotated ]]; then
            python $BIN/annotate.py $snpeff $ANNOTATION | gzip -c > $annotated
        fi
    done

    # create a vcf of normal intersection
    filtered_vcfs=""
    for (( i = 0; i < ${#normal_samples[@]}; i++ )); do
        normal=${normal_samples[$i]}
        vcf=$RESULTS/$normal/$normal.freebayes.vcf
        filtered=$RESULTS/$normal/$normal.filtered.freebayes.vcf
        filtered_vcfs=$filtered" "$filtered_vcfs
        if [[ ! -f $filtered ]]; then
            cat $vcf | java -jar $SNPSIFT filter "((QUAL >= 500) & (DP >= 100))" | java -jar $SNPSIFT intervals $TARGETS > $filtered
        fi
    done

    # combine the normal VCFs into a single
    if [[ ! -f $NORMALVCF ]]; then
        cat $filtered_vcfs >> tmp.vcf
        bedtools sort -i tmp.vcf > $NORMALVCF
        rm tmp.vcf
    fi

    # run all samples against normal database
    for (( i = 0; i < ${#tumor_samples[@]}; i++ )); do
        tumor=${tumor_samples[$i]}
        vcf=$intersect_results/$tumor.vcf
        snpeff=$intersect_results/$tumor.snpeff.txt.gz
        annotated=$intersect_results/$tumor.annotated_snps.txt.gz
        if [[ ! -f $vcf ]]; then
            bedtools intersect -v -a $RESULTS/$tumor/$tumor.freebayes.vcf -b $NORMALVCF > $vcf
        fi
        if [[ ! -f $snpeff ]]; then
            java -jar $SNPEFF eff -noStats -v -c $SNPEFFCONFIG \
                -fi $TARGETS -o txt -minQ 100 -minC 10 -no-downstream -no-intergenic \
                -no-intron -no-upstream canFam3 $vcf | gzip -c > $snpeff
        fi
        if [[ ! -f $annotated ]]; then
            python $BIN/annotate.py $snpeff $ANNOTATION | gzip -c > $annotated
        fi
    done
</code></pre>

</div>

</body>
</html>